crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,337 @@
+    pub fn storage_arc(mut self, storage_arc: Arc) -> Self {
+        self.storage_arc = storage_arc;
+        self
+    }
+
+    /// what is this agent's uptime
+    /// - 0 = never reachable
+    /// - 1 = always reachable
+    pub fn uptime_0_to_1(mut self, uptime_0_to_1: f64) -> Self {
+        self.uptime_0_to_1 = uptime_0_to_1;
+        self
+    }
+}
+
+/// builder for a set of peers we know about
+/// this should generally either be our entire storage arc of peer info
+/// or, if it is huge, small sample arc
+pub struct RValuePeerRecordSet {
+    arc_of_included_peer_records: Arc,
+    peer_records: Vec<RValuePeerRecord>,
+}
+
+impl Default for RValuePeerRecordSet {
+    fn default() -> Self {
+        Self {
+            arc_of_included_peer_records: Arc::new(0.into(), 0),
+            peer_records: Vec::new(),
+        }
+    }
+}
+
+impl RValuePeerRecordSet {
+    /// the arc we are representing with this PeerRecordSet
+    pub fn arc_of_included_peer_records(mut self, arc_of_included_peer_records: Arc) -> Self {
+        self.arc_of_included_peer_records = arc_of_included_peer_records;
+        self
+    }
+
+    /// push a new PeerRecord into this record set
+    pub fn push_peer_record(mut self, peer_record: RValuePeerRecord) -> Self {
+        if !self
+            .arc_of_included_peer_records
+            .contains_location(peer_record.storage_arc.center())
+        {
+            panic!("peer record does not fit within arc_of_included_peer_records");
+        }
+        self.peer_records.push(peer_record);
+        self
+    }
+}
+
+/// Interpolate qualified redundancy in the network given our current
+/// knowledge of peers' claimed storage arcs.
+/// For a given storage arc (should match *this* agent's storage arc),
+/// if all the peers we know about are actually storing what they claim to be
+/// we can interpolate what the whole network "R" value is
+/// also taking into account our experience of those agents' reachability.
+pub fn interpolate_r_value_for_given_arc(peer_record_set: &RValuePeerRecordSet) -> f64 {
+    let mut running_pct_total: f64 = 0.0;
+
+    for record in peer_record_set.peer_records.iter() {
+        // first get the percentage of total space this peer is covering
+        let mut relative_pct_of_space_covered: f64 =
+            record.storage_arc.length() as f64 / ARC_LENGTH_MAX as f64;
+
+        // if their uptime is less than 1,
+        // they only count directly proportional to their uptime
+        relative_pct_of_space_covered *= record.uptime_0_to_1;
+
+        // update our running percentage covered
+        running_pct_total += relative_pct_of_space_covered;
+    }
+
+    let pct_of_space_covered: f64 =
+        peer_record_set.arc_of_included_peer_records.length() as f64 / ARC_LENGTH_MAX as f64;
+
+    // we are only sampling the pct_of_space_covered
+    // we need to grow our sample (interpolate) to the rest of the space
+    running_pct_total * (1.0 / pct_of_space_covered)
```

Comment:
Let's say we've got two buckets of equal size (half the space), bucket A is known, and bucket B is unknown.

Bucket A has a single item in it that covers 1/3 of the TOTAL space.

If we interpolate such that we assume bucket B has an identical item that also covers 1/3 of the total space, we should assume that 2/3 of the total space is covered, and this math checks out:

```
(1 / 3) * (1 / 0.5) == (2 / 3)
```

Thus, we do want to know the total percentage coverage for the items within our small sample, and then multiply for assumed counterparts in the unknown portion of the circle.

Reply:
Considering the storage arc not as "storage" but "the portion of the arc for which I know agents," the current logic makes sense to me. Basically, @neonphog is calculating just how much "expected availability" of data across the entire DHT is provided by the agents within that arc. Interpolation by scaling up WRT the portion of the DHT the arc covers should provide a valid estimate of the "expected availability" of data in the entire DHT.

I sense a clipping approach could provide useful insight, too, but would require much more complex math to interpolate a valid DHT-wide result.

Reply:
Also, I'd suggest replacing the term `pct` with something akin to `DHT_coverage` ... aggregate expected coverage beyond 100% is quite likely here (consider an agent with a storage arc > .5 * DHT_size), which feels funny to conceptualize as a "percent"

Reply:
Hehe, indeed, if you look at my unit-tests, they are looking for an "R" value between 25 and 50... aka 2500% to 5000% : )

Reply:
The problem I see: there is no guarantee that all `peer_records` are completely included in the `arc_of_included_peer_records`. If one is not, the calculation is off -> solution: clipping `record.storage_arc` with `arc_of_included_peer_records` (i.e. just calculate the intersection of those arcs).

The alternative would be to regard this constraint as part of the contract of this function. There is no problem if it is never used on data that does not adhere to this implicit contract. I believe that the cost of adding above mentioned clipping is far less then potential debugging in the future.

Reply:
My understanding of the intended rrDHT sharding approach is that it requires an agent to maintain awareness of ALL active peers within its storage arc. While there may be short periods where it doesn't know all peers, the impact on the calculation of expected coverage should be negligible... but @neonphog may have updated this intention since I last had a deep dive?

Reply:
@lucksus - Let me try to re-phrase your comment - please tell me if I'm not hearing you correctly.

Let's say we have an arc with only 10 unit positions within it. That is, your "Location" can be 0-9 inclusive. Given the same equal sized A and B buckets above, let's define A as 0-4 inclusive, and B as 5-9. That is they, both have 5 unit length.

Now, our `arc_of_included_peer_records` is A (0-4). The peer we know about is located at location 0, with a radius of 2, so, the start of its arc is actually 9, and the inclusive end of its arc is 1. (its length is 4 exclusive).

That 9 position is outside our `arc_of_included_peer_records` so you are suggesting we clip the arc to match, which would cause us to treat it as if it were covering 0 to 1.

---

I submit that we don't actually care about the specific positioning of these agent's arcs. We are dealing in aggregate here, and while some of the arcs we know about may extend beyond our range, it is also true that some of the arcs we don't know about (because their center location is beyond our storage arc) actually extend into our range as well.

If we were to clip our entries to our arc, it would actually make our calculations less accurate because we wouldn't be taking this second case into account.

Reply:
Another way to get the same result might be to:
- step 1 - figure out the average total coverage for all known nodes.
- step 2 - guess how many total nodes there are based on our known node count.
- step 3 - guess the total coverage by multiplying steps 1 and 2.

This method, I think, makes it extra clear that we wouldn't want to clip, because then our average coverage would be incorrect. I *think* what's currently implemented gives the same result, but perhaps it would be better to switch to this for code clarity.

Reply:
@jmday / @lucksus - removed pct references and implemented alternate calculation strategy. Please provide feedback.